### ISSUE TITLE
New Module "Restaurant Kitchen Screens" [$100]

### ISSUE THREAD

**Issue by yelizariev** [itpp-labs/pos-addons#197]
Dear Odoo users, developers and business experts,
 
we are going to create new module *Restaurant Kitchen Screens*, which allows to see orders and report on readiness via display at a kitchen. We believe that it will be useful for Restaurants. It will have LGPL-3 licence and will be based on our module [Sync restaurant orders](https://github.com/it-projects-llc/pos-addons/tree/10.0/pos_multi_session_restaurant).

Before we start, we would like to know your opinions, ideas, suggestions. Especially, we are gathering for information about typical restaurant workflows, feature requirements.

For sponsoring and specific customisation concerns as well as for private discussion email us directly at pos@it-projects.info

Also, this development can be sponsored via bountysource service (see links below)

**UPD 12 June:** 
Plan for initial release is ready: https://github.com/it-projects-llc/pos-addons/issues/197#issuecomment-314087554 . We hope to crowdfund at least 1000 US$. You are welcome to support [updates](https://github.com/it-projects-llc/pos-addons/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20pos_multi_session%20OR%20pos_multi_session_restaurant%20) for base modules *[pos_multi_session](https://www.odoo.com/apps/modules/10.0/pos_multi_session/)* and *[pos_multi_session_restaurant](https://www.odoo.com/apps/modules/10.0/pos_multi_session_restaurant/)* too.

We will add all sponsors to README file of the new module (name + link or email).

@multiwebinc -- 100$ (bounty)
@W360S -- 100$ (direct payment)

**Comment by dvdhinesh** [comment]
As you know, a typical restaurant will vary from the normal retail showroom POS. Given that, we will discuss a couple of use cases.

Use case **Pizza Shop**

This Pizza shop sells pizza in slices, whole etc and they make slices without payment internally.
 
**Flow 1:** Take order at cash counter -> Send to Kitchen -> Kitchen Display shows that we have a new order -> They produce it (Estimated time to finish entered manually and sent to cash counter) -> Kitchen order done -> Cash counter display shows that it was finished on Kitchen -> They bill it -> Orders done.
**Flow 2:** Take order at cash counter -> They bill it first -> Then the rest are as above.

**Note:** It would be best if we have an option for communicating directly from the Kitchen to Cash counter in case if any product like toppings goes out of order and they want to provide an alternative which requires a prior customer approval.

**Q1:** How can we manage the order change? All the products are maintained as consumables in this shop. Do we provide run time order change until it was paid?
**Q2:** How can we manage internal orders without payment that goes to Kitchen? Consider this shop is selling slices and they want to prepare it without payment.

Odoo's Bill print will not work without the proxy for kitchen print, so we have to render a receipt for the browser.
I have made an addon [POS Kitchen ticket](https://github.com/dvdhinesh/pos-addons-repo/tree/8.0/pos_kitchen_ticket) to simplify those things to some extent.

Use case **Advanced Restaurant**

This is an Advanced restaurant with multiple kitchens and may be a separate fruit juice/beverages bay. They will place a device on the table and the customer can order by themselves. Based on the product categories, the orders needs to split and sent to different screens whereas the main Cash counter screen needs to look at the whole order at payment (maybe review it before sending it to other screens).

For the customer screen, we may not need the payment options while we need those for cash counter.

Here the customer screen and the rest needs to communicate with each other if they want to give an alternative and change the ongoing orders.

I hope this could be helpful.

**PS:** I haven't reviewed that [Sync restaurant order](https://github.com/it-projects-llc/pos-addons/tree/10.0/pos_multi_session_restaurant) module and gone deep to the POS workings yet. The above is purely a functional requirement that I got from a friend of mine who is a Business Analyst.

**Comment by yelizariev** [comment]
Thank you @dvdhinesh for information.

First of all, as there could be a special screen where customer can order by themselves, it's better either to name module **Restaurant Screens** or make two different modules. We will need to think about it. 

I think that instead of implementing some specific workflow, we can create customizable interface. Draft solution is below.

-----------

It could be based on concept "State" of a product (orderline):

* each orderline has as many custom "states" as it's needed. Examples of states:

  * time to finish cooking
  * confirmed by user or not
  * enough products or not
  * Need a Sault (or any other ingredient) or not
  * Priority
  * etc.
  * as well as simple text note
  
* On order screen there could be added buttons with custom text and icon that change one or several state. Buttons are shown on some condition (e.g. product category or states value)
* On order screen updated states are displayed (each value of the state can has own color)
* Kitchen screen similarly shows buttons and current states

-----------

сс @ilmir-k

**Comment by dvdhinesh** [comment]
@yelizariev 
:+1: For the generic **Restaurant Screens**  and **Customizable Interface.** 

We could add a boolean to the POS setup screen (like Virtual keyboard, scrollbars etc) for the default screen like whether it is going to serve Kitchen, Cash counter or maybe Customer.

The overall draft solution seems interesting.

**Comment by mslokhat** [comment]
Hi,
some ideas from openbravopos ( i used this long time ago) : 
http://forge.openbravo.com/plugins/mwiki/index.php/Openbravoposmonitor/Instructions

**Comment by mslokhat** [comment]
seems already exists :
https://apps.openerp.com/apps/modules/9.0/pos_restaurant_kitchen/

**Comment by yelizariev** [comment]
@mslokhat the module pos_restaurant_kitchen is not opensource. Also, it's based on modified copies of modules from this repository (they don't respect neither LGPL licence nor authorship)

**Comment by mslokhat** [comment]
Oups, didn't know : maybe to contact Odoo for this
i'm in for developping something for the kitchen

1. Bar and coffee Screen
ex : drinks has to be prepared by the barman and should be sent to a screen in the bar
waiter should be informed when the drink is ready

2. stage for preparation : sent, ready, taken by Waiter

3. Stats for kitchen (duration, ....)

4. Self service POS : 
on the table Tablet with self service pos with Ads
how to deal with customer who plays with order and refuse to pay after delivery saying they did not order it

5. Chat with kitchen screen (2 ways)
6. instructions : should be kept internally

**Comment by mslokhat** [comment]
you know this : http://www.acode.com/en/?p=322

**Comment by berpweb** [comment]
Very eager to play with this module. This screen thing is a lag for Odoo POS for many years.
Let us know when you're done with it.

**Comment by okaki** [comment]
Any follow up on this? Very interested :)

**Comment by yelizariev** [comment]
It's postponed for a while as our resources are overloaded at the moment. Thanks for your interest

**Comment by makobec** [comment]
please notify me if ever you finished it

**Comment by ovnicraft** [comment]
@yelizariev we can fund the development if you have the time.

**Comment by 360org** [comment]
yeap i can help with the idea of requirement :) i have work & control the hotel and restaurant before.

**Comment by yelizariev** [comment]
@W360S please share your thoughts as I still not sure how to do it better.

**Comment by yelizariev** [comment]
@ovnicraft do you have your view of how it should be implemented? 
As for sponsoring, it should definitely speed it up. You can [post a bounty](https://www.bountysource.com/issues/39896947-new-module-restaurant-kitchen-screens?utm_campaign=plugin&utm_content=tracker%2F2289114&utm_medium=issues&utm_source=github) to share funding with other interested people. For other options contact us by email pos@it-projects.info

**Comment by 360org** [comment]
ok i will give you all detail restaurant workflow and you will understand easily

**Comment by 360org** [comment]
Hello here is your fully restaurant workflow: https://drive.google.com/file/d/0B2v96YMQkx-BVmVCZHdhTll5LU0/view?usp=sharing
you can chek the file or check the image it same 
![restaurant workflow 3](https://user-images.githubusercontent.com/8590556/27783566-e676f4a0-6000-11e7-829d-3652d1a15843.png)

**Comment by 360org** [comment]
also i see your multi restaurant  sync section need some fixed to make it work easy to client i may open new case.

**Comment by 360org** [comment]
if you not clear anything can ask me

**Comment by yelizariev** [comment]
Ok, we have a plan for initial release and then we can discuss further updates.

Initial release

* Settings at pos.config:
  * checkbox "Kitchen Screen"
  * Product categories for this kitchen (similar to "Printed Product Categories" at Order printer)
  * checkbox "show Floors plan" -- if set, user at the kitchen will see floor plan first and then see orders at that table. Otherwise user sees all orders
  * checkbox "show lines from other kitchens" -- user will be able to see other lines, but will not able to update their status. Otherwise user see only lines for his kitchens

* Order line status can be tracked per each quantity update, e.g.

  * Cap of Coffee - 4 items:

    * 2 coffees - Delivered to table
    * 2 coffees - Confirmed

* Used time will be show between orderline status, e.g.

  * Pizzas 3 items

    * 2 pizza: Confirmed - 0:30 -  Cooking started - 14:33 - Cooked - 1:00 - Delivered to table
    * 1 pizza: Confirmed - 0:42 -  Cooking started - 9:01 (active timer) 

**UPD 12 june:**

Custom States and buttons

* pos.order.line.state

  * Display Name
  * Technical name
  * Type: Stage or Tag

    * Stage: Something like "confirmed", "started", "done". Time between stages is accounted
    * Tag: attach information to order line (e.g. "No Ice", "No sugar" etc.)

* pos.order.line.button

  * name -- button label
  * color -- button color
  * condition_code -- js code to check shall button be visible. Following variables can be used:
      * is_kitchen
      * state -- current state
      * product
      * quantity
      * price
      * order_is_paid
      * ...
  * action_state -- state to apply on clicking
  * action_close  -- does close order. Order will be removed from all screens

**Comment by yelizariev** [comment]
@multiwebinc,
is it ok, if your bounty will cover only updates I described [above](https://github.com/it-projects-llc/pos-addons/issues/197#issuecomment-314087554)?

**Comment by 360org** [comment]
have sent you little sponsor :) hope that we can finish that task early :) goodluck

**Comment by mslokhat** [comment]
Hi,
some questions :
. does the cooker can update the status (either using on screen button or keyboard shorcuts)
. does the order automatically disappear when all items have been delivered or ticket closed (ex : paid on the cashier) (did you see how McDo works)
. does it support multi kitchens (ex : bar screen, cold kitchen, ..)
. reports and stats : duration of an order , ...

i'll be able to sponsor surely USD 100 , maybe 200.
rgds

**Comment by multiwebinc** [comment]
@yelizariev Use it as you wish. You guys have made a lot of great add-ons over the years, so I'm happy to give a little bit back. This is just something that I was personally interested in having implemented in my restaurant, so I thought I would support it.

My only small suggestion based on your list would be make the order line status items be able to be enabled/disabled. A smaller restaurant may only care about whether an order is confirmed and when it's done and not care about adding extra work to cooks/servers having to select all the different in-between stages.

Other than that, you might want to look at examples of how other POS systems handle kitchen displays and pick one you think works best for a wide range of restaurant types and emulate that. However it seems to me you know what you're doing.

**Comment by yelizariev** [comment]
@W360S I don't see you payment at this moment. I'll check it again later.

**Comment by yelizariev** [comment]
@multiwebinc thanks for your help and review. I think we need to return to custom states idea ( https://github.com/it-projects-llc/pos-addons/issues/197#issuecomment-267027525 ). I'll come back with more details. I need to think how to make it flexible and not too complex.

**Comment by 360org** [comment]
@yelizariev yeap i believe that still spending will become soon

**Comment by yelizariev** [comment]
@mslokhat 

> . does the cooker can update the status (either using on screen button or keyboard shorcuts)

Yes, cooker will be able to update status, but I don't think to add keyboard shortcuts support to first release plan.

> . does the order automatically disappear when all items have been delivered or ticket closed (ex : paid on the cashier) (did you see how McDo works)

Initially I planned to close order only after payment, but now I recognize, that at this case we cannot support workflows when payment is done before order is closed. I'll add it to the [plan](https://github.com/it-projects-llc/pos-addons/issues/197#issuecomment-314087554)

> . does it support multi kitchens (ex : bar screen, cold kitchen, ..)

yes

> . reports and stats : duration o

I plan to store this information to pos.order, but without reports and stats for the first time.

**Comment by yelizariev** [comment]
ok, here is a draft plan for custom order line states



Custom States and buttons

* pos.order.line.state

  * Display Name
  * Technical name
  * Type: Stage or Tag

    * Stage: Something like "confirmed", "started", "done". Time between stages is accounted
    * Tag: attach information to order line (e.g. "No Ice", "No sugar" etc.)

* pos.order.line.button

  * name -- button label
  * color -- button color
  * condition_code -- js code to check shall button be visible. Following variables can be used:
      * is_kitchen
      * state -- current state
      * product
      * quantity
      * price
      * order_is_paid
      * ...
  * action_state -- state to apply on clicking
  * action_close  -- does close order. Order will be removed from all screens

**Comment by yelizariev** [comment]
@W360S I confirm your 100$ payment. Thank you!

**Comment by 360org** [comment]
Welcome :)

**Comment by berpweb** [comment]
@yelizariev We would like to enhance the Odoo POS for various business cases.
We would like you to develop 2 modules and open source it to the community.

1) Adding customer signature on the POS order. Very similar to https://apps.odoo.com/apps/modules/10.0/pos_signature/

2) Adding customer photo from web cam on the POS order for credit sales.

We would like to fund the development or crowd source it.
Please let us know the procedure for processing further...

**Comment by yelizariev** [comment]
@berpweb first of all, you need to create a separate issue per each task that can be done independently

**Comment by 360org** [comment]
Hey @yelizariev any news for kitchen screen project?

**Comment by yelizariev** [comment]
@W360S I made updates in first [message](https://github.com/it-projects-llc/pos-addons/issues/197#issue-194572644) and forgot to notify about it.
So, we hope to fundraise about 1000$ for this module or will start it anyway when we will have time for it. Sorry that we cannot start it soon.

**Comment by 360org** [comment]
ok, that good @yelizariev I think it will be starting soon :)
Hi, all! I can see that we have more people here if you all raise 1 hand it will be done soon :) we have 200$ just 800$ more
@dvdhinesh @mslokhat @berpweb @okaki 
@makobec @ovnicraft

**Comment by randall-vx** [comment]
Hi @yelizariev, hope you are ok !

Please check this workflow, I was used few months ago when we implemented a POS in a little Brew Pub.

![pos-workflow](https://user-images.githubusercontent.com/16024775/29388717-82d3485c-82a3-11e7-8bc8-4c5f40dad5a2.png)

**Comment by alaherma** [comment]
Can i get notified when this module is done?

**Comment by yelizariev** [comment]
@alaherma I will post here on any updates

**Comment by DjeblounSidali** [comment]
Any news ?

**Comment by Raoufasker** [comment]
How much time is left to finish this module?

**Comment by ovnicraft** [comment]
@yelizariev i cant find at Bountysource the budget required to get this module done, there is any definition ?

**Comment by yelizariev** [comment]
@ovnicraft the required budget is specified at the first message of this issue. I am not sure how to do it at Bountysource.com

**Comment by yelizariev** [comment]
Some news.

One developer (I suppose his profile at github is @dhecar) made the module on top of pos_multi_session by himself. The module will be published on apps store only (license of pos_multi_session allows to do it). I have a private access to review it. If the implementation is good enough, I'll retire from making alternative solution, but will sponsor this issue to be implemented by someone else.

**Comment by ovnicraft** [comment]
@yelizariev sounds great, we are ready to fund or buy module.

**Comment by 360org** [comment]
Thanks to @dhecar we dying to see @yelizariev  :)

**Comment by domedfd** [comment]
Hello, it is possible to buy the module and what price.
sorry I do not speak English!

**Comment by 360org** [comment]
You can sponsors as you can

Thanks & Best Regards
Chau Le’s
World 360° Services
Sent from iPhone

On Jan 11, 2018, at 08:06, Domenico Delvalle <notifications@github.com<mailto:notifications@github.com>> wrote:


Hello, it is possible to buy the module and what price.
sorry I do not speak English!

**Comment by GabbasovDinar** [comment]
Hello all

I started work on the issue. You can check the progress of the work here: https://github.com/it-projects-llc/addons-dev/pull/374
I wait for your feedback and suggestions for improving the module.

Thanks!

**Comment by gustavovalverde** [comment]
@yelizariev, what amount are you needing to speed up this Feature? 😄

**Comment by chafarleston** [comment]
where I can contribute. paypal account or something @yelizariev,

**Comment by yelizariev** [comment]
@chafarleston you can fund the development here: https://www.bountysource.com/issues/39896947-new-module-restaurant-kitchen-screens
(works via paypal)

**Comment by Bhaniramka** [comment]
I am a end user. Requirement is for a cub house where multiple kitchens and other facilities are there. Members use prepaid magnetic stripe cards to avail any facility. We are planning to upgrade the system since current system doesnt have kitchen and token display system in place. Current software was custom designed.

Looking for a similar thing right now where app and UI can be customised as per need. App will be hosted on a local server and clients will be connected on intranet.

Pls let me know if this it is possible? Customizing UI is where i am stuck. We can pay for the services.

I will be available on a_bhaniramka@yahoo.com to discuss further

**Comment by adduzz** [comment]
@yelizariev Would appreciate to have a look at this module, let me if we can connect for a quick demo 
ping me on skype: adnan.dev or adnan.dev@gmail.com

**Comment by Ramil-Mukhametzyanov** [comment]
@adduzz 
Hello, I am going to prepare demo on Monday

**Comment by adduzz** [comment]
@ramil-mukhametzyanov
Looking forward to it

**Comment by yelizariev** [comment]
I'd like to inform all participants that some work was done and published here: https://github.com/it-projects-llc/addons-dev/tree/pos-addons-10.0-pos_multi_session_kitchen/pos_multi_session_kitchen

Also, we decided to drop  pos_multi_session support since odoo 13.0, cause there is built-in pos order syncronization. So, I'm not sure about the future of the Kitchen Screens module. If you want to discuss it, ping us by email pos@it-projects.info

**Comment by multiwebinc** [comment]
@yelizariev Thank you for the update. I had completely forgot that I posted a bounty on this. I also no longer use Odoo since we subleased our restaurant out and the tenants use a different POS that their bank provided them with. Since the work was not finished, I am not sure if I should accept or reject the bounty. If you feel you still deserve it, please let me know.

**Comment by yelizariev** [comment]
@multiwebinc I'm not sure about bounty either :)
I have similar changes on my side. I'm out of business for developing new modules and working on maintaining modules instead. The work that was done belongs to @it-projects-llc represented by @ilmir-k. And it's better ask their point of view.

**Comment by chafarleston** [comment]
@yelizariev for odoo 11 no there
